provider/telegram.go

```diff
@@ -0,0 +1,395 @@
+package provider
+
+import (
+	"context"
+	"crypto/sha1"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	neturl "net/url"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/dgrijalva/jwt-go"
+	"github.com/go-pkgz/auth/logger"
+	authtoken "github.com/go-pkgz/auth/token"
+	"github.com/go-pkgz/rest"
+	"github.com/pkg/errors"
+)
+
+// TelegramHandler implements login via telegram
+type TelegramHandler struct {
+	logger.L
+
+	ProviderName         string
+	ErrorMsg, SuccessMsg string
+
+	TokenService TokenService
+	AvatarSaver  AvatarSaver
+	Telegram     TelegramAPI
+
+	mu           sync.Mutex                 // Guard for the map below
+	authRequests map[string]authRequestInfo // Tokens waiting for confirmation
+}
+
+// TelegramAPI is used for interacting with telegram API
+type TelegramAPI interface {
+	GetUpdates(ctx context.Context) (*telegramUpdate, error)
+	Avatar(ctx context.Context, userID int) (string, error)
+	Send(ctx context.Context, id int, text string) error
+}
+
+type authRequestInfo struct {
+	confirmed bool // whether login request has been confirmed and userInfo set
+	expires   time.Time
+	user      *userInfo
+}
+
+type userInfo struct {
+	ID     int
+	Name   string
+	Avatar string
+}
+
+// changed in tests
+var pollInterval = time.Second
+
+// Run starts processing login requests sent in Telegram
+// Blocks caller
+func (t *TelegramHandler) Run(ctx context.Context) error {
```

Comment:
I've named the method `Run` because of it's blocking, long-running nature. The same thing as with `func (c *CustomServer) Run(ctx context.Context)`. Isn't that valid?

Reply:
fine with me